File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # BFS
        # T: O(N)
        # S: O(M)
        levelOrderArray, q = list(), [root]
        
        if not root:
            return root
        
        while q:
            nodeCount = len(q)
            levelValues = list()
            
            while nodeCount > 0:
                node = q.pop(0)
                levelValues.append(node.val)
                
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
                
                nodeCount -= 1
                
            levelOrderArray.append(levelValues)
            
        return levelOrderArray
        
        
        
        
            
            

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        /*
            Algorithms & DS: #BFS, #Recursion
            Time: O(N), Extra Space: O(N)
        */
        List<List<Integer>> ans = new ArrayList<>();
        levelOrder(root, 1, ans);
        
        return ans;
    }
    private void levelOrder(TreeNode node, int depth, List<List<Integer>> ans){
        if(node == null){
            return;
        }
        
        while(ans.size() < depth){
            ans.add(new ArrayList<>());
        }
        
        ans.get(depth - 1).add(node.val);
        levelOrder(node.left, depth + 1, ans);
        levelOrder(node.right, depth + 1, ans);
    }
}